File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>>ans;
    vector<int>cnt;
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        dfs(0,candidates,target);
        return ans;
    }
    void dfs(int start,vector<int>& candidates, int target){
        if(target==0){ans.push_back(cnt);return;}
        for(int i=start;i<candidates.size();i++){
                if(i>start && candidates[i]==candidates[i-1])continue;
                int x = candidates[i];
                if(x>target)break;
                cnt.push_back(x);
                dfs(i+1,candidates,target-x);
                cnt.pop_back();
        }
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
     candidates.sort()
     res = []
     path = []
     def bt(start,remain):
        if remain==0:
            res.append(path[:])
            return
        if remain<0:
            return 
        for i in range(start,len(candidates)):
            if i>start and candidates[i]==candidates[i-1]:
                 continue
            path.append(candidates[i])
            bt(i+1,remain-candidates[i])
            path.pop()
     bt(0,target)
     return res